### PR TITLE
Offering extended configuration

### PR DESCRIPTION
**Pull request by ajs6f**
Adds configuration for use with [Trellis-Cassandra](https://github.com/ajs6f/trellis-cassandra).

### CODE REVIEW

**Comment by acoburn** [comment]
This seems fine, but I wonder if it wouldn't be better for the trellis-over-cassandra-app to subclass `TrellisConfiguration` and add these methods there?

**Comment by ajs6f** [comment]
You tell me-- I know as much about Dropwizard as I do about playing poker: nothing. Should I close this and try that instead?

**Comment by ajs6f** [comment]
Or do I just cast the config? Is it legit to:
```

```

**Comment by ajs6f** [comment]
Sorry-- misclick.

**Comment by acoburn** [comment]
A better approach might be to add a `@JsonAnySetter` method to `TrellisConfiguration`:

```java
@JsonAnySetter
public void set(final String name, final String value) {
    internalMap.put(name, value);
}

@JsonAnyGetter
public Map<String, String> any() {
    return internalMap;
}
```

That adds a simple extension point for any possible subclasses.

**Comment by ajs6f** [comment]
Where is `internalMap`? Is that to be added as well?

**Comment by ajs6f** [comment]
I'm not much with JSON; what about `Map<String, Object>`? Would that effectively account for nesting?

**Comment by acoburn** [comment]
I haven't ever used that precise construct with Jackson serializations, but I don't see why it wouldn't work. It should be trivial to set up a test -- just add some additional fields to one of the configuration files in `./src/test/resources`

**Comment by ajs6f** [comment]
Didn't work at all, not even with simple `String`s:
```
io.dropwizard.configuration.ConfigurationParsingException: /Users/ajs6f/Documents/trellis/trellis/trellis-app/bin/test/config1.yml has an error:
  * Unrecognized field at: cassandraAddress
    Did you mean?:
      - assets
      - resources
      - namespaces
      - cacheMaxAge
      - binaries
        [12 more]
```

So maybe that works for Jackson's semantics, but not for Dropwizard's particular concrete processing code?

**Comment by ajs6f** [comment]
Again, I know little about Dropwizard, but I'm beginning to suspect that we're running up against its intentional limits-- it seems to be for building single apps, not ecosystems. Extensible config is really not a concern if you are build a single app.

**Comment by ajs6f** [comment]
Okay, how's that?

**Comment by ajs6f** [comment]
incidentally, I tried going:
```
public class TrellisApplication extends Application<TrellisConfiguration> {

```
=>
```
public class TrellisApplication<T extends TrellisConfiguration> extends Application<T> {

```
but all of the test classes exploded with:
```
The constructor DropwizardTestSupport<TrellisConfiguration>(Class<TrellisApplication>, String, ConfigOverride, ConfigOverride, ConfigOverride) is undefined
```

Not really sure what's going on there, but I _think_ the issue is that `Class<TrellisApplication>` isn't provably a supertype of `Class<TrellisApplication<T extends TrellisConfiguration>>` (or the other way 'round), so the compiler pukes into its own shirt. Shall we just rewrite the whole app in Scala?

**Comment by christopher-johnson** [comment]
@ajs6f I tried running the test and it works for me using IntelliJ.  One side note is that the `@RunWith(JUnitPlatform.class)` is not required since the test class is written for JUnit5.  AFAIK, that annotation invokes the JUnit4 runner.

**Comment by ajs6f** [comment]
Thanks much @christopher-johnson ! I didn't actually write that test, just added a few limes to it, so I'll let @acoburn remove that annotation. Unless, @acoburn , you'd like me to do it here?

File: trellis-app/src/main/java/org/trellisldp/app/config/TrellisConfiguration.java [hunk 1]
```diff
@@ -13,10 +13,18 @@
...
 
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 import io.dropwizard.Configuration;
 
+import static java.util.Collections.synchronizedMap;
+
+import java.util.Collections;
```
Comment (acoburn) [inline trellis-app/src/main/java/org/trellisldp/app/config/TrellisConfiguration.java]: This isn't used and can be removed.

File: trellis-app/src/main/java/org/trellisldp/app/config/TrellisConfiguration.java [hunk 2]
```diff
@@ -13,10 +13,18 @@
  */
 package org.trellisldp.app.config;
 
+import com.fasterxml.jackson.annotation.JsonAnyGetter;
+import com.fasterxml.jackson.annotation.JsonAnySetter;
 import com.fasterxml.jackson.annotation.JsonProperty;
 
 import io.dropwizard.Configuration;
 
+import static java.util.Collections.synchronizedMap;
```
Comment (acoburn) [inline trellis-app/src/main/java/org/trellisldp/app/config/TrellisConfiguration.java]: Move `static import`s to the top (before ordinary imports)

File: trellis-app/src/main/java/org/trellisldp/app/config/TrellisConfiguration.java [hunk 3]
```diff
@@ -62,6 +70,8 @@
     private String baseUrl = null;
 
     private String resourceLocation = null;
+
```
Comment (acoburn) [inline trellis-app/src/main/java/org/trellisldp/app/config/TrellisConfiguration.java]: trailing whitespace

File: trellis-app/src/test/java/org/trellisldp/app/config/TrellisConfigurationTest.java [hunk 4]
```diff
@@ -49,6 +52,14 @@ public void testConfigurationGeneral1() throws Exception {
         assertNull(config.getResources());
         assertEquals((Integer) 2, config.getBinaryHierarchyLevels());
         assertEquals((Integer) 1, config.getBinaryHierarchyLength());
+        assertEquals("my.cluster.node", config.any().get("cassandraAddress"));
+        assertEquals((Integer)245993, config.any().get("cassandraPort"));
+        @SuppressWarnings("unchecked")
+        Map<String, Object> extraConfig = (Map<String, Object>) config.any().get("extraConfigValues");
```
Comment (acoburn) [inline trellis-app/src/test/java/org/trellisldp/app/config/TrellisConfigurationTest.java]: add `final` here.

File: trellis-app/src/test/java/org/trellisldp/app/config/TrellisConfigurationTest.java [hunk 5]
```diff
@@ -49,6 +52,14 @@ public void testConfigurationGeneral1() throws Exception {
...
+        assertTrue((Boolean) extraConfig.get("one"));
+        @SuppressWarnings("unchecked")
+        List<String> list = (List<String>) extraConfig.get("two");
```
Comment (acoburn) [inline trellis-app/src/test/java/org/trellisldp/app/config/TrellisConfigurationTest.java]: add `final`

**Comment by ajs6f** [comment]
I will fix those probs in two seconds, but just so you know-- I ran `./gradlew check install` to check this and it passed fine. Do we need to couple `checkstyleMain` into `check`? I thought it already was?

**Comment by acoburn** [comment]
That's odd. It could be that running this without `clean` means that the changes aren't caught. I definitely get the checkstyle errors when running `./gradlew clean check install`

**Comment by ajs6f** [comment]
Okay, I'll keep an eye peeled. I might try making a few intentional faux pas and see which Gradle invocations catch them and which do not.

**Pull request merged**